src/lib/ledger_catchup/ledger_catchup.ml

```diff
@@ -1,31 +1,124 @@
 open Pipe_lib
 open Coda_base
 
-module type Inputs_intf = sig
-  include Transition_frontier.Inputs_intf
-
-  module Transition_frontier :
-    Transition_frontier_intf
-    with type state_hash := State_hash.t
-     and type external_transition := External_transition.t
-     and type ledger_database := Ledger.Db.t
-     and type staged_ledger := Staged_ledger.t
-     and type masked_ledger := Ledger.Mask.Attached.t
-     and type transaction_snark_scan_state := Staged_ledger.Scan_state.t
-end
-
-module Make (Inputs : Inputs_intf) :
+module Make (Inputs : Inputs.S) :
   Catchup_intf
   with type external_transition := Inputs.External_transition.t
    and type transition_frontier := Inputs.Transition_frontier.t
    and type transition_frontier_breadcrumb :=
               Inputs.Transition_frontier.Breadcrumb.t
-   and type state_hash := State_hash.t = struct
-  let run ~frontier:_ ~catchup_job_reader ~catchup_breadcrumbs_writer:_ =
-    don't_wait_for
-      (Strict_pipe.Reader.iter catchup_job_reader ~f:(fun _ ->
-           failwith "Intentionally unimplemented catchup" ))
+   and type state_hash := State_hash.t
+   and type network := Inputs.Network.t = struct
+  open Inputs
+  module Worker = Catchup_worker.Make (Inputs)
+
+  let construct_breadcrumb_path ~logger initial_staged_ledger
+      external_transitions =
+    let open Or_error.Let_syntax in
+    let%map _, breadcrumbs =
+      List.fold_result external_transitions ~init:(initial_staged_ledger, [])
+        ~f:(fun (staged_ledger, acc) transition_with_hash ->
+          let external_transition = With_hash.data transition_with_hash in
+          let diff =
+            External_transition.staged_ledger_diff external_transition
+          in
+          let%map _, _, `Updated_staged_ledger staged_ledger =
+            Staged_ledger.apply ~logger staged_ledger diff
+          in
+          let new_breadcrumb =
+            Transition_frontier.Breadcrumb.create transition_with_hash
+              staged_ledger
+          in
+          (staged_ledger, new_breadcrumb :: acc) )
+    in
+    List.rev breadcrumbs
+
+  let materilize_breadcrumbs ~frontier ~logger = function
+    | [] -> `Success []
+    | root_transition :: _ as external_transitions -> (
+        let initial_state_hash =
+          With_hash.data root_transition
+          |> External_transition.protocol_state
+          |> External_transition.Protocol_state.previous_state_hash
+        in
+        match Transition_frontier.find frontier initial_state_hash with
+        | None -> `No_matching_root_hash initial_state_hash
+        | Some initial_breadcrumb -> (
+            let initial_staged_ledger =
+              Transition_frontier.Breadcrumb.staged_ledger initial_breadcrumb
+            in
+            match
+              construct_breadcrumb_path ~logger initial_staged_ledger
+                external_transitions
+            with
+            | Ok result -> `Success result
+            | Error e -> `Error e ) )
```

Comment:
This would be cleaner if you used the `Or_error` monad instead of an anonymous ADT (you could still make the error cases unique by using an anonymous ADT for the error type parameter of `Or_error.t`).